### PR TITLE
Add unit tests for scansourcefiles actor

### PR DESCRIPTION
**Pull request by tomasfratrik**
Test if files are correctly tracked by actor. Create mocked existent/non-existent, modified/non-modified, rpm owned/not-owned files. Also test non-existent version files.

Jira: OAMG-10367

### CODE REVIEW

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:49]: You could store `api.produce.model_instances[0].files[INDEX_OF_VERSION_FILE]` inside a variable instread to avoid a lot of the repetition below.

**Review by matejmatuska — CHANGES_REQUESTED** [review]
It is a common practice to first test the lower level functions as units and then mock them when testing functions that call them.

So in all of these 3 tests, you would just test the functionality on  `scansourcefiles.scan_file()`. Then you would add a test that tests `scansourcefiles.process()` and mock `scan_file()` in it, because that function is already covered. 

This would reduce the amount of mocks in the lower level tests and also allow you to run the test once per possible input (in this case a file). Then you could use `@pytest.mark.parametrize()` to run the test for multiple inputs (files). You can also specify the expected results there to avoid having specific asserts for each one of them.

You can take a look at https://github.com/oamg/leapp-repository/blob/master/repos/system_upgrade/el8toel9/actors/rocescanner/tests/unit_test_rocescanner.py for an example of such tests. First `get_roce_nics_lines` is properly tested and then it's mocked when testing `process`.

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (tomasfratrik) [reply repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:17]: I was using `os.path.join` from start, but thought It would probably be more slower (because code uses many instances of it). But you are right, it is overally better, readibility, cross-platform ...

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (tomasfratrik) [reply repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:49]: Was hesitant about this, because didn't see other actors doing this (didn't check enough of them). But yes, there is no reason for this repetition, which makes it worse

Comment (matejmatuska) [reply repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:49]: It isn't really an issue, but since you already have variable for the index you might as well store the object since you don't need the index anywhere else.

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (matejmatuska) [reply repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:17]: I think it would require a lot more calls to notice any performance hit, anyway the test already run for quite a while and in this case correctness > performance.

Comment (pirat89) [reply repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:17]: @tomasfratrik I recommend you to watch the following video to understand when it's good and when it's premature to optimise something: https://youtu.be/tKbV6BpH-C8 

In short, optimisation that is faster by 0.01s does not make sense when the execution time is in minutes.

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (matejmatuska) [reply repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:89]: Also please add a case for rpm owned file both modified and unmodified.

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:119]: Mock the scan_files function directly and return what it is expected to return. You can do that because that one is already covered by the test above.

Also, you shouldn't need to mock the TRACKED_FILES, since all calls operating directly on files can be mocked, e.g. the os.path.exists call (as explained in one of my other comments).

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:148]: The same applies as described in the comment in `test_rpm_owned_files`. In this case it is valid to mock TRACKED_FILES as there are no files by version in the actor library, but again it is not required for the `file_not_rpm_owned` to physically exists (see the other comments).

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:134]: This name is not very descriptive, please choose a more descriptive name. But there is nothing wrong with not extracting this into a variable if it's properly formatted in the parametrize.

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:43]: Nothing really wrong with this, but you could use basically any path such as `/use/somefile` as long as it's valid. There is not need for it to be at the path you used.

**Review by matejmatuska — CHANGES_REQUESTED** [review]
It's now looking a lot better. However some additional changes are requested.

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (pirat89) [inline repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:102]: that's kind of duplicate input data and also prevents testing combination of files that exists or not, that are owned by rpm or not. The first line specifies params that must be found in all later defined files + names of files.  You can obtain the same result with params defined on the first line, or just with the last param. The second way is what I would do, as It could also introduce testing for various combinations. Another benefit is that input data are more readable.

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (pirat89) [inline repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:11]: why? what is value in comparison to do just this?

```suggestion
RHEL_MAJOR_VERSIONS_LIST = [8, 9]
```

Also, it's incorrect as `get_source_major_version` returns string, not integer. so correctly it would be:
```suggestion
RHEL_MAJOR_VERSIONS_LIST = ['8', '9']
```

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (pirat89) [inline repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py:112]: this is nice example that there is no reason for duplicated input data

**Review by pirat89 — CHANGES_REQUESTED** [review]

**Comment by tomasfratrik** [comment]
My notes:

* I separated some values in pytest parametrize, because I thinkit is now more readable (i can see it faster what is what, and it is not too much longer).
* I split test of function `_get_rpm_name` into 2 tests, 1 containing normal tests, and other one  error. Now it could considered as 'redundant' (even though the code still differs), but I think it is more readable.
* My thought process over test test_tracked_files, which tests process, was that, I will give 'you' (process function), tracked files, and major version, and you do the rest. Process should just take files from 'common' and 'major_version' but I "don't know that", so I let it do whatever it wants to, and then compare. And mainly compare produced message.

I will squash commits, and write proper commit message after approved, or when it is finalized.

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (tomasfratrik) [reply repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py]: Done.

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (tomasfratrik) [reply repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py]: Done.

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (tomasfratrik) [reply repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py]: Yes I see that now, done.

File: repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py
Comment (tomasfratrik) [reply repos/system_upgrade/common/actors/scansourcefiles/tests/unit_test_scansourcefiles.py]: Done.

**Review by pirat89 — APPROVED** [review]
that's much better! lgtm!

**Comment by tomasfratrik** [comment]
When splitting `_get_rpm_name` I forgot to check warnmsg, should be fixed now.

**Comment by pirat89** [comment]
@fernflower seems that automatic building is broken by general.

**Review by pirat89 — APPROVED** [review]

**Review by matejmatuska — APPROVED** [review]

**Pull request merged**
